index.bs

```diff
@@ -101,21 +118,190 @@ spec: webidl
+
+    <div algorithm="request-permission-to-use">
+      <p>
+        To <dfn export>request permission to use</dfn> a |descriptor|, the UA
+        must perform the following steps. Note that these steps may wait for
+        user input, so they should not be used from other algorithms running on
+        the main thread. This algorithm returns either {{"granted"}} or
+        {{"denied"}}.
+      </p>
+      <ol>
+        <li>
+          If |descriptor|'s <a>permission state</a> is not {{"prompt"}}, return
+          that value and abort these steps.
+        </li>
+        <li>
+          Show the user a prompt asking their permission for the calling
+          algorithm to use powerful feature described by |descriptor|.
+        </li>
+        <li>
+          If the user grants permission, return {{"granted"}}; otherwise return
+          {{"denied"}}. Depending on the details of the user's interaction, the
+          UA may also treat this as new information about the user's intent for
+          other realms with the same origin.
+
+          <p class="note">
+            This is intentionally vague about the details of the permission UI
+            and how the UA infers user intent. UAs should be able to explore
+            lots of UI within this framework.
+          </p>
+        </li>
+      </ol>
+    </div>
+
+    <div algorithm="prompt-user-to-choose">
+      <p>
+        To <dfn export>prompt the user to choose</dfn> one of several options
+        associated with a |descriptor|, the UA must perform the following steps.
+        Note that these steps may wait for user input, so they should not be
+        used from other algorithms running on the main thread. This algorithm
+        returns either {{"denied"}} or one of the options.
+      </p>
+      <ol>
+        <li>
+          If |descriptor|'s <a>permission state</a> is {{"denied"}},
+          return {{"denied"}} and abort these steps.
+        </li>
+        <li>
+          If |descriptor|'s <a>permission state</a> is {{"granted"}}, the UA may
+          return one of the options and abort these steps. If the UA returns
+          without prompting, then successive <a lt="prompt the user to
+          choose">prompts for the user to choose</a> from the same set of
+          options with the same |descriptor| should return the same option,
+          unless the UA receives new information about the user's intent.
+        </li>
+        <li>
+          Show the user a prompt asking them to choose one of the options or
+          deny permission, and wait for them to choose. If the calling algorithm
+          specified extra information to show in the prompt, show it.
+        </li>
+        <li>
+          If the user chose an option, return it; otherwise return {{"denied"}}.
+          Depending on the details of the user's interaction, the UA may also
+          treat this as new information about the user's intent for other realms
+          with the same origin.
```

Comment:
So, given what is written here, it would cause a propagation across realms. It sounds like "queue a task" might be needed for propagating the change, no?